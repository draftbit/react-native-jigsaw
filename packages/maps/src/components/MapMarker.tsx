import * as React from "react";
import { Marker } from "./ReactNativeMaps";
import { MapMarkerProps } from "@draftbit/types";

const MapMarker: React.FC<React.PropsWithChildren<MapMarkerProps>> = ({
  latitude,
  longitude,
  title,
  description,
  pinColor,
  flat,
  style,
  children,
}) => {
  if (!Marker) {
    return null;
  }

  return (
    <Marker
      coordinate={{
        latitude,
        longitude,
      }}
      title={title != null ? String(title) : undefined}
      description={description != null ? String(description) : undefined}
      flat={flat}
      pinColor={pinColor}
      style={style}
    >
      {children}
    </Marker>
  );
};

export default MapMarker;
