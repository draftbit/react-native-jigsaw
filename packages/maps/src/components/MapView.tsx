import * as React from "react";
import NativeMapView, { Marker } from "./ReactNativeMaps";
import { MapViewProps } from "@draftbit/types";

// Dynamically import from ./ReactNativeMaps so that we don't
// require react-native-maps unless we're in native.

// Approximates a conversion of zoom level to altitude,
// since different platforms require different values.
// https://stackoverflow.com/a/37142662
function zoomToAltitude(zoom: number) {
  const A = 40487.57;
  const B = 0.00007096758;
  const C = 91610.74;
  const D = -40467.74;

  return C * Math.pow((A - D) / (zoom - D) - 1, 1 / B);
}

class MapView extends React.Component<MapViewProps> {
  private mapRef: React.RefObject<any>;
  constructor(props: MapViewProps) {
    super(props);
    this.state = {};
    this.mapRef = React.createRef();
  }

  componentDidUpdate(prevProps: MapViewProps) {
    if (
      prevProps.latitude != null &&
      prevProps.longitude != null &&
      this.props.latitude != null &&
      this.props.longitude != null &&
      (prevProps.latitude !== this.props.latitude ||
        prevProps.longitude !== this.props.longitude)
    ) {
      this.animateToLocation({
        latitude: this.props.latitude,
        longitude: this.props.longitude,
        zoom: this.props.zoom,
      });
    }
  }

  animateToLocation({
    latitude,
    longitude,
    zoom,
  }: {
    latitude: number;
    longitude: number;
    zoom?: number;
  }) {
    const args: {
      center: { latitude: number; longitude: number };
      altitude?: number;
      zoom?: number;
    } = {
      center: {
        latitude,
        longitude,
      },
    };

    if (zoom) {
      args.altitude = zoomToAltitude(zoom || 1);
      args.zoom = zoom;
    }

    this.mapRef.current.animateCamera(args);
  }

  render() {
    const {
      provider,
      latitude,
      longitude,
      zoom,
      showsCompass = false,
      rotateEnabled = true,
      zoomEnabled = true,
      loadingEnabled = true,
      scrollEnabled = true,
      loadingBackgroundColor,
      loadingIndicatorColor,
      mapType = "standard",
      showsUserLocation,
      followsUserLocation,
      showsPointsOfInterest,
      style,
      children,
    } = this.props;

    if (!NativeMapView || !Marker) {
      return null;
    }

    const camera = {
      altitude: zoomToAltitude(zoom || 1),
      heading: 0,
      pitch: 0,
      zoom,
      center: {
        latitude,
        longitude,
      },
    };

    return (
      <NativeMapView
        ref={this.mapRef}
        provider={provider}
        mapType={mapType}
        showsCompass={showsCompass}
        rotateEnabled={rotateEnabled}
        zoomEnabled={zoomEnabled}
        camera={camera}
        loadingEnabled={loadingEnabled}
        scrollEnabled={scrollEnabled}
        showsUserLocation={showsUserLocation}
        followsUserLocation={followsUserLocation}
        showsPointsOfInterest={showsPointsOfInterest}
        loadingBackgroundColor={loadingBackgroundColor}
        loadingIndicatorColor={loadingIndicatorColor}
        style={style}
      >
        {children}
      </NativeMapView>
    );
  }
}

export default MapView;
