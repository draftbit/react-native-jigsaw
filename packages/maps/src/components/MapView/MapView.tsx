import * as React from "react";
import NativeMapView from "./ReactNativeMaps";
import { MapViewProps } from "@draftbit/types";

// Dynamically import from ./ReactNativeMaps so that we don't
// require react-native-maps unless we're in native.

const MapView: React.FC<MapViewProps> = ({
  provider,
  latitude,
  latitudeDelta,
  longitude,
  longitudeDelta,
  showsCompass = false,
  rotateEnabled = true,
  zoomEnabled = true,
  loadingEnabled = true,
  scrollEnabled = true,
  loadingBackgroundColor,
  loadingIndicatorColor,
  mapType = "standard",
  style,
  children,
}) => {
  if (!NativeMapView) {
    return null;
  }

  return (
    <NativeMapView
      provider={provider}
      mapType={mapType}
      showsCompass={showsCompass}
      rotateEnabled={rotateEnabled}
      zoomEnabled={zoomEnabled}
      initialRegion={{
        latitude: latitude as number,
        longitude: longitude as number,
        latitudeDelta: latitudeDelta as number,
        longitudeDelta: longitudeDelta as number,
      }}
      loadingEnabled={loadingEnabled}
      scrollEnabled={scrollEnabled}
      loadingBackgroundColor={loadingBackgroundColor}
      loadingIndicatorColor={loadingIndicatorColor}
      style={style}
    >
      {children}
    </NativeMapView>
  );
};

export default MapView;
