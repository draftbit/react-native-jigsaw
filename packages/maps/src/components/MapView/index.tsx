// The native version of the Map components. @draftbit/native is used by
// @draftbit/ui, React Native package that includes web and native experiences.
// This is here so that we don't need to install react-native-maps in the
// version of Jigsaw that's a dependency of the builder (@draftbit/ui).
export { default as MapView } from "./MapView";
export { default as MapMarker } from "./MapMarker";
export { default as MapCallout } from "./MapCallout";
