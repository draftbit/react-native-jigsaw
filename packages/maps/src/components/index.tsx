export { default as MapCallout } from "./MapCallout";
export { default as MapMarker } from "./MapMarker";
export { default as MapView } from "./MapView";
