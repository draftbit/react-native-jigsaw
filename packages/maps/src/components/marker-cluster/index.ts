export { default as MapMarkerCluster } from "./MapMarkerCluster";
export { default as MapMarkerClusterView } from "./MapMarkerClusterView";
