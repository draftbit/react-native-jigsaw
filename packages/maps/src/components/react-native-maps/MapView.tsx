export { default as default } from "react-native-maps";
