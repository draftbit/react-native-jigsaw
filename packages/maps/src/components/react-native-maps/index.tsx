export { default as default } from "./MapView";
export { Callout } from "./Callout";
export { Marker } from "./Marker";
export { Circle } from "./Circle";
