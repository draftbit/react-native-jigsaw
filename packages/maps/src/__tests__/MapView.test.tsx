import React from "react";
import { act, render, screen } from "@testing-library/react-native";
import MapView from "../components/MapView";
import MapMarker from "../components/MapMarker";
import { MapMarkerCluster } from "../components/marker-cluster";
import { Marker as MapMarkerComponent } from "../components/react-native-maps";
import { MarkerClusterer } from "@teovilla/react-native-web-maps";

const mockAnimateCamera = jest.fn();

jest.mock("../components/react-native-maps", () => {
  const React = require("react");

  class MapView extends React.Component {
    render(): React.ReactNode {
      return <>{this.props.children}</>;
    }

    animateCamera = mockAnimateCamera;
  }

  class Marker extends React.Component {
    render(): React.ReactNode {
      return <>{this.props.children}</>;
    }
  }

  return {
    __esModule: true,
    default: MapView,
    Marker,
  };
});

jest.mock("@teovilla/react-native-web-maps", () => {
  const MarkerClusterer = (props: any) => {
    return <>{props.children}</>;
  };

  return {
    MarkerClusterer,
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

describe("MapView tests", () => {
  test("should animateToLocation update map location", () => {
    const ref = React.createRef<MapView<any>>();

    render(<MapView ref={ref} apiKey="" />);

    act(() => {
      ref.current?.animateToLocation({
        latitude: 40.741895,
        longitude: -73.989308,
        zoom: 10,
      });
    });

    expect(mockAnimateCamera.mock.lastCall).toMatchInlineSnapshot(`
      [
        {
          "altitude": 86346.53718958268,
          "center": {
            "latitude": 40.741895,
            "longitude": -73.989308,
          },
          "heading": 0,
          "pitch": 0,
          "zoom": 10,
        },
      ]
    `);
  });

  test("should render markers provided as children", () => {
    render(
      <MapView apiKey="">
        <MapMarker key={1} latitude={41.741895} longitude={-73.989308} />
        <MapMarker key={2} latitude={42.741895} longitude={-73.989308} />
        <MapMarker key={3} latitude={43.741895} longitude={-73.989308} />
        <MapMarker key={4} latitude={44.741895} longitude={-73.989308} />
      </MapView>
    );

    const renderedMarkers = screen
      .UNSAFE_queryAllByType(MapMarkerComponent)
      .map((marker) => marker.props.coordinate);

    expect(renderedMarkers).toMatchInlineSnapshot(`
      [
        {
          "latitude": 41.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 42.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 43.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 44.741895,
          "longitude": -73.989308,
        },
      ]
    `);
  });

  test("should render markers provided as data using renderItem", () => {
    const markers = [
      { latitude: 41.741895, longitude: -73.989308 },
      { latitude: 42.741895, longitude: -73.989308 },
      { latitude: 43.741895, longitude: -73.989308 },
      { latitude: 44.741895, longitude: -73.989308 },
    ];

    render(
      <MapView
        markersData={markers}
        renderItem={({ item }) => (
          <MapMarker latitude={item.latitude} longitude={item.longitude} />
        )}
        apiKey=""
      />
    );

    const renderedMarkers = screen
      .UNSAFE_queryAllByType(MapMarkerComponent)
      .map((marker) => marker.props.coordinate);

    expect(renderedMarkers).toMatchInlineSnapshot(`
      [
        {
          "latitude": 41.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 42.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 43.741895,
          "longitude": -73.989308,
        },
        {
          "latitude": 44.741895,
          "longitude": -73.989308,
        },
      ]
    `);
  });

  test("should render manually created marker clusters", () => {
    render(
      <MapView apiKey="">
        <MapMarkerCluster>
          <MapMarker latitude={41.741895} longitude={-73.989308} />
          <MapMarker latitude={42.741895} longitude={-73.989308} />
        </MapMarkerCluster>

        <MapMarkerCluster>
          <MapMarker latitude={43.741895} longitude={-73.989308} />
          <MapMarker latitude={44.741895} longitude={-73.989308} />
        </MapMarkerCluster>
      </MapView>
    );

    const renderedClusters = screen
      .UNSAFE_queryAllByType(MarkerClusterer)
      .map((marker) => marker.props.children);

    expect(renderedClusters).toMatchInlineSnapshot(`
      [
        [
          <Marker
            coordinate={
              {
                "latitude": 41.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
          <Marker
            coordinate={
              {
                "latitude": 42.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
        ],
        [
          <Marker
            coordinate={
              {
                "latitude": 43.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
          <Marker
            coordinate={
              {
                "latitude": 44.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
        ],
      ]
    `);
  });

  test("should automatically create clusters from markers when autoClusterMarkers is true", () => {
    render(
      <MapView apiKey="" autoClusterMarkers>
        <MapMarker latitude={40.741895} longitude={-73.989308} />
        <MapMarker latitude={40.741895} longitude={-73.979308} />

        <MapMarker latitude={43.741895} longitude={-73.989308} />
        <MapMarker latitude={43.741895} longitude={-73.979308} />
      </MapView>
    );

    const renderedClusters = screen
      .UNSAFE_queryAllByType(MarkerClusterer)
      .map((marker) => marker.props.children);

    expect(renderedClusters).toMatchInlineSnapshot(`
      [
        [
          <Marker
            coordinate={
              {
                "latitude": 40.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
          <Marker
            coordinate={
              {
                "latitude": 40.741895,
                "longitude": -73.979308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
        ],
        [
          <Marker
            coordinate={
              {
                "latitude": 43.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
          <Marker
            coordinate={
              {
                "latitude": 43.741895,
                "longitude": -73.979308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
        ],
      ]
    `);
  });

  test("should MapMarker render when wrapped in fragment", () => {
    render(
      <MapView apiKey="">
        <>
          <MapMarker latitude={40.741895} longitude={-73.989308} />
        </>
      </MapView>
    );

    const renderedMarkers = screen
      .UNSAFE_queryAllByType(MapMarkerComponent)
      .map((marker) => marker.props.coordinate);

    expect(renderedMarkers).toMatchInlineSnapshot(`
      [
        {
          "latitude": 40.741895,
          "longitude": -73.989308,
        },
      ]
    `);
  });

  test("should MapMarkerCluster render when wrapped in fragment", () => {
    render(
      <MapView apiKey="">
        <>
          <MapMarkerCluster>
            <MapMarker latitude={41.741895} longitude={-73.989308} />
            <MapMarker latitude={42.741895} longitude={-73.989308} />
          </MapMarkerCluster>
        </>
      </MapView>
    );

    const renderedClusters = screen
      .UNSAFE_queryAllByType(MarkerClusterer)
      .map((marker) => marker.props.children);

    expect(renderedClusters).toMatchInlineSnapshot(`
      [
        [
          <Marker
            coordinate={
              {
                "latitude": 41.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
          <Marker
            coordinate={
              {
                "latitude": 42.741895,
                "longitude": -73.989308,
              }
            }
            onPress={[Function]}
            tracksViewChanges={true}
          />,
        ],
      ]
    `);
  });
});
