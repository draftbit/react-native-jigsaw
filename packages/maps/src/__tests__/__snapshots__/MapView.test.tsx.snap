// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView tests should MapMarker render when wrapped in fragment 1`] = `1`;

exports[`MapView tests should MapMarkerCluster render when wrapped in fragment 1`] = `1`;

exports[`MapView tests should animateToLocation update map location 1`] = `
[
  {
    "altitude": 86346.53718958268,
    "center": {
      "latitude": 40.741895,
      "longitude": -73.989308,
    },
    "heading": 0,
    "pitch": 0,
    "zoom": 10,
  },
]
`;

exports[`MapView tests should automatically create clusters from markers when autoClusterMarkers is true 1`] = `2`;

exports[`MapView tests should render manually created marker clusters 1`] = `2`;

exports[`MapView tests should render markers provided as children 1`] = `
[
  {
    "latitude": 41.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 42.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 43.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 44.741895,
    "longitude": -73.989308,
  },
]
`;

exports[`MapView tests should render markers provided as data using renderItem 1`] = `
[
  {
    "latitude": 41.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 42.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 43.741895,
    "longitude": -73.989308,
  },
  {
    "latitude": 44.741895,
    "longitude": -73.989308,
  },
]
`;
